### DISCUSSION TITLE
fastapi-users and Jinja2 - how do I create a custom /login, /logout, /register , etc.

### DISCUSSION THREAD

**Discussion by HENRYHKll** [fastapi-users/fastapi-users#1409] [Q&A] [answered]
Hi, I'm new to Fastapi.  I use a bunch of fastapi + Jinja2, in order to register a new user or log in or reset my password, I use forms. how do I create a custom /login , etc?
**transport:** cookie
**strategy:**  jwt 

This is my prototype router, tried to do it this way, but failed
```py
@router.post("/login")
async def login(
    user_manager: BaseUserManager = Depends(get_user_manager),
    cooki: AuthenticationBackend = Depends(auth_backend.get_strategy),
    password: str = Form(...),
    email: str = Form(...),
):
    credentials = OAuth2PasswordRequestForm(username=email, password=password)
    user = await user_manager.authenticate(credentials)
    if user is None:
        print("Unauthorized")
    else:
        print("Authorized")
```
I always have a user = None. Please help me, maybe I'm thinking in the wrong direction

**Comment by hgalytoby** [comment]
I am using the example https://fastapi-users.github.io/fastapi-users/latest/configuration/full-example/

I tested it with your login method and there is no problem.

You should attach more information.

<img width="1233" alt="image" src="https://github.com/user-attachments/assets/d83a4a37-3fc9-4444-ba77-ac781cd2748e">

**Reply by HENRYHKll** [reply]
Hi, thanks for responding and checking the code. I discovered the problem, I was inattentive.
https://github.com/HENRYHKll/fastapi-users-jinja2/blob/91a0dabe7a86564b33af934674852bb4238ec30a/app/app.py#L142
Do you know why logout doesn't work in this code?

**Reply by hgalytoby** [reply]
When you say it doesn't work, do you mean the token is still active after logging out?

https://fastapi-users.github.io/fastapi-users/latest/configuration/authentication/strategies/jwt/#logout

**Reply by HENRYHKll** [reply]
No get_current_user_token does not return a tuple, but returns a reference to a function

**Reply by hgalytoby** [reply, marked as answer]
```py
def get_current_user_token():
    user_token = user_authenticator.current_user_token(active=True, verified=False)
    return user_token


@app.post("/logout")
async def logout_post(
    user_token: Authenticator = Depends(get_current_user_token()),
    strategy: Strategy = Depends(auth_backend.get_strategy),
):
    user, token = user_token
    print(user.email, user.id, token)
    return await auth_backend.logout(strategy, user, token)
```

or

```py
get_current_user_token = user_authenticator.current_user_token(active=True, verified=False)


@app.post("/logout")
async def logout_post(
    user_token: Authenticator = Depends(get_current_user_token),
    strategy: Strategy = Depends(auth_backend.get_strategy),
):
    user, token = user_token
    print(user.email, user.id, token)
    return await auth_backend.logout(strategy, user, token)
```

**Reply by HENRYHKll** [reply]
Thank you so much for helping me!
